File: data_structures/02_linked_lists/01_reverse_linked_list/solution.rb
class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, next_node = nil)
    @val = val
    @next = next_node
  end
end

def reverse_list(head)
  prev = nil
  current = head

  while !current.nil?
    next_node = current.next
    current.next = prev
    prev = current
    current = next_node
  end

  prev
end

File: data_structures/02_linked_lists/01_reverse_linked_list/solution.js
class ListNode {
  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

function reverseList(head) {
  let prev = null;
  let current = head;

  while (current !== null) {
    const nextNode = current.next;
    current.next = prev;
    prev = current;
    current = nextNode;
  }

  return prev;
}

module.exports = { ListNode, reverseList };
